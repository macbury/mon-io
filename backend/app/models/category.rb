class Category < ApplicationRecord
  include Kindable

  belongs_to :user, optional: true

  has_many :transactions, dependent: :destroy
  has_many :receipts, dependent: :destroy
  has_many :imports, dependent: :destroy
  has_many :category_location_suggestions, dependent: :destroy
  has_many :category_budgets, dependent: :destroy

  validates :name, presence: true, uniqueness: { case_sensitive: false }
  validates :color, presence: true
  validates :icon, presence: true

  scope :for_user, ->(user) { where('shared = true OR user_id IS NULL OR user_id = :user_id', user_id: user.id) }
  scope :not_archived, -> { where(archived: false) }
  scope :for_ids, ->(category_ids) { category_ids.empty? ? all : where('id NOT IN (?)', category_ids) }
  scope :expense_or_tax, -> { where({ kind: [Category.kinds[:expense], Category.kinds[:tax]] }) }
  scope :income_or_saving, -> { where({ kind: [Category.kinds[:income], Category.kinds[:saving], Category.kinds[:loan]] }) }

  def currency
    local_currency = self[:currency]
    if local_currency
      Money::Currency.new(local_currency)
    else
      Money.default_currency
    end
  end
end
