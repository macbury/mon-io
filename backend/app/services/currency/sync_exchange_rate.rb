module Currency
  class SyncExchangeRate < Service
    def call
      CurrencyRate.sync!

      currencies.each do |currency|
        rate = find_rate(main_currency, currency)
        ExchangeRateService.add_rate(main_currency, currency, rate, today) if rate

        next_rate = find_rate(currency, main_currency)
        ExchangeRateService.add_rate(currency, main_currency, next_rate, today) if next_rate
      end

      convert_metals_ounces_into_grams

      old_exchange_rates
        .where(base: main_currency)
        .where('currency NOT IN (?)', used_currencies)
        .delete_all

      old_exchange_rates
        .where('base NOT IN (?)', used_currencies)
        .where(currency: main_currency)
        .delete_all
    end

    private

    GRAM_TO_OUNCE = 0.03215 # 1 Gram Weight = 0.03215 Troy Ounce

    METALS = {
      'XAU' => 'GAU',
      'XAG' => 'GAG'
    }.freeze

    def convert_metals_ounces_into_grams
      METALS.each do |from, to|
        rate = Money.new('1', from).exchange_to(main_currency) * GRAM_TO_OUNCE
        ExchangeRateService.add_rate(to, main_currency, rate.to_f, today)
        ExchangeRateService.add_rate(main_currency, to, (1.0 / rate.to_f), today)
      end
    end

    def old_exchange_rates
      ExchangeRate.where('date <= ?', 1.month.ago)
    end

    def used_currencies
      @used_currencies ||= Transaction.group(:amount_currency).pluck(:amount_currency)
    end

    def today
      @today ||= Time.zone.today
    end

    def main_currency
      Money.default_currency.iso_code.upcase
    end

    def currencies
      Money::Currency.all.map(&:to_s)
    end

    def find_rate(from, to)
      CurrencyRate.fetch_fiat(from, to) || CurrencyRate.fetch_crypto('Coinbase', from, to)
    end
  end
end