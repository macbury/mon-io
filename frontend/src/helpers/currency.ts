import { useMemo } from 'react'
import { Money } from '../api/graphql'

export function formatMoney(money : Money, abs = false, round = false, minimumFractionDigits?: number) {
  if (!money) {
    return '-'
  }

  const amount = abs ? Math.abs(money.cents) : money.cents

  if (!minimumFractionDigits) {
    minimumFractionDigits = money.currency.subunitToUnit.toString().split('').length - 1
  }

  if (round) {
    return Math.round(amount / money.currency.subunitToUnit).toLocaleString(undefined, { minimumFractionDigits }) + ' ' + money.currency.symbol
  } else {
    return (amount / money.currency.subunitToUnit).toLocaleString(undefined, { minimumFractionDigits }) + ' ' + money.currency.symbol
  }
}

export function moneyParts(money : Money) {
  if (money && money.currency) {
    let [dollars, cents] = (money.cents / money.currency.subunitToUnit).toPrecision().split('.')
    if (!cents) {
      cents = '00'
    }
    if (cents.length == 1) {
      cents = `${cents}0`
    }
    return [parseInt(dollars), cents, money.currency.symbol]
  } else {
    return [0, 0, 'EUR']
  }
}

export function useFormatMoney(money : Money, abs = false, round = false, minimumFractionDigits?: number) {
  return useMemo(() => formatMoney(money, abs, round, minimumFractionDigits), [money, abs, round, minimumFractionDigits])
}

export function useMoneyParts(money : Money) {
  return useMemo(() => moneyParts(money), [money])
}