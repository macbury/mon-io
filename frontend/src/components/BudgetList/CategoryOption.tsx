import React, { useCallback } from 'react'
import { TouchableRipple, Text } from 'react-native-paper'
import styled from 'styled-components/native'
import CategoryIcon from '../CategoryIcon'
import { Category, Money, TransactionCategoryKind } from '../../api/graphql'
import { useFormatMoney } from '../../helpers/currency'

interface ICategoryNameProps {
  color: string;
}

export interface ICategoryOptionProps {
  category: Category
  spend: Money
  width: number
  type: TransactionCategoryKind
  onCategoryPress(category : Category, type: TransactionCategoryKind)
}

const CategoryName = styled(Text)`
  color: ${(props : ICategoryNameProps) => props.color};
  text-align: center;
  margin-bottom: 10px;
`

const CurrentlySpend = styled(Text)`
  color: ${(props : ICategoryNameProps) => props.color};
  text-align: center;
  margin-bottom: 10px;
`

interface IContainerProps {
  width?: number;
  present?: boolean;
}

const ItemContainer = styled(TouchableRipple)`
  width: ${(props : IContainerProps) => `${props.width}%`};
  height: 118px;
  padding: 2px;
  padding-top: 12px;
  border-radius: 10px;
  overflow: hidden;
`

const Icon = styled(CategoryIcon)`
  opacity: ${({ present } : IContainerProps) => present ? 1.0 : 0.5 };
`

const InnerContainer = styled.View`
  display: flex;
  justify-content: center;
  align-content: center;
  align-items: center;
`

export default function CategoryOption(props : ICategoryOptionProps) {
  const { type, spend, category, width, onCategoryPress } = props
  const kind = category.kind === TransactionCategoryKind.Tax ? category.kind : type
  const onPress = useCallback(() => onCategoryPress(category, kind), [onCategoryPress, category, kind])
  const amount = useFormatMoney(spend)
  const present = spend.cents !== 0

  return (
    <ItemContainer
      width={width}
      onPress={onPress}
      rippleColor={`${category.color}3C`}>
      <InnerContainer>
        <CategoryName color={category.color} numberOfLines={1}>{category.name}</CategoryName>
        <Icon
          present={present}
          size={40}
          color={category.color}
          name={category.icon} />
        <CurrentlySpend color={category.color}>{present ? amount : '-'}</CurrentlySpend>
      </InnerContainer>
    </ItemContainer>
  )
}