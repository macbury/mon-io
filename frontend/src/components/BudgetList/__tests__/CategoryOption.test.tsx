import React from 'react'
import { shallow } from 'enzyme'
import snapshot from 'enzyme-to-json'
import { category } from '../../../test/factories/category'
import { currency } from '../../../test/factories/currency'

import CategoryOption, { ICategoryOptionProps } from '../CategoryOption'

it('renders', () => {
  const props : ICategoryOptionProps = {
    category: category.build(),
    onCategoryPress: () => ({}),
    width: 60,
    spend: {
      cents: 0,
      currency: currency.build()
    }
  }

  const component = shallow(<CategoryOption {...props}/>)
  expect(snapshot(component)).toMatchSnapshot()
})