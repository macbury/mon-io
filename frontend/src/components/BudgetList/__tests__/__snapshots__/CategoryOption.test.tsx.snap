// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<Styled(withTheme(TouchableRipple))
  onPress={[Function]}
  rippleColor="#fff3C"
  width={60}
>
  <Styled(View)>
    <Styled(withTheme(Text))
      color="#fff"
      numberOfLines={1}
    >
      Category
    </Styled(withTheme(Text))>
    <Styled(CategoryIcon)
      color="#fff"
      name="MaterialIcons:repeat"
      present={false}
      size={40}
    />
    <Styled(withTheme(Text))
      color="#fff"
    >
      -
    </Styled(withTheme(Text))>
  </Styled(View)>
</Styled(withTheme(TouchableRipple))>
`;
